File: C/Archivos/ARCH_ESTAD_2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Ventas 
{
   int id;   
   char tipo[7]; //C: Contado, T: Cuenta Corriente
   int fecha; //aaaammdd   
   char cliente[50];   
   int vendedor; //1..15
   float monto;
};

//No se usa mas por que uso una sola instruccion para esto...
void PONER_A_CERO_MATRIZ(float m[3][16]) 
{
  int i, j;

  for(i = 0; i < 2; ++i) 
  {
    for(j = 0; j < 15; ++j) 
    {
      m[i][j] = 0;      
    }
  }
}

void MOSTRAR_MATRIZ(float m[3][16]) 
{
  int i, j;

  printf("\n---------------------\n");
  printf("Resultados: \n");

  for(i = 0; i < 3; ++i) 
  {    
    for(j = 0; j < 16; ++j) 
    {
      printf("%3.2f\t", m[i][j]); //Formateo la salida real...
    }
    //Dejo un renglon por cada fila
    printf("\n");
  }
}


int main () 
{
    FILE *ent;
    struct Ventas reg;
    float estad[3][16] = {{0}}; //Defino e inicializo en cero toda la matriz
    int i, j;

    if ((ent = fopen ("datos.dat", "r")) == NULL) 
    {
      fprintf(stderr, "\nError al abrir el archivo de entrada...\n");
      exit (1);
    }

    //Matriz para acumular - Se pone a cero
    //PONER_A_CERO_MATRIZ(estad);

    while(fread(&reg, sizeof(reg), 1, ent) == 1)  // No Fin de Archivo
    {
      printf ("%d  %s  %d  %s  %d  %f\n", reg.id, reg.tipo, reg.fecha, reg.cliente, reg.vendedor, reg.monto);      

      //Mapeo tipo de venta
      if(strcmp(reg.tipo, "Contado")==0)
        i = 0;
      else
        i = 1;
      

      //Mapeo vendedor
      j = reg.vendedor - 1;

      //Acumulo por Tipo y Vendedor
      estad[i][j] += reg.monto;  //estad[i][j] = estad[i][j] + reg.monto;

      estad[2][j] += reg.monto; //Total por Vendedor
      estad[i][15] += reg.monto; //Total por Tipo
      estad[2][15] += reg.monto;  //Total general
    }
    
    fclose (ent);

    //Muestro resultados antes de salir
    MOSTRAR_MATRIZ(estad);

    return 0;
}

File: Python/Archivos/ARCH_ESTAD_2.py
import sys
import struct

def mostrar_matriz(m):
    print("\n---------------------")
    print("Resultados: ")
    for i in range(len(m)):
        for j in range(len(m[i])):
            print(f"{m[i][j]:3.2f}\t", end='')
        print()

class Ventas:
    def __init__(self, id_, tipo, fecha, cliente, vendedor, monto):
        self.id = id_
        self.tipo = tipo
        self.fecha = fecha
        self.cliente = cliente
        self.vendedor = vendedor
        self.monto = monto

def main():
    estad = [[0.0]*16 for _ in range(3)]
    try:
        with open("datos.dat", "rb") as ent:
            while True:
                data = ent.read(4 + 7 + 4 + 50 + 4 + 4)
                if not data or len(data) < 4+7+4+50+4+4:
                    break
                id_, tipo_bytes, fecha, cliente_bytes, vendedor, monto = struct.unpack('i7s i 50s i f', data)
                tipo = tipo_bytes.decode('utf-8').rstrip('\x00')
                cliente = cliente_bytes.decode('utf-8').rstrip('\x00')
                reg = Ventas(id_, tipo, fecha, cliente, vendedor, monto)
                print(f"{reg.id}  {reg.tipo}  {reg.fecha}  {reg.cliente}  {reg.vendedor}  {reg.monto}")
                i = 0 if reg.tipo == "Contado" else 1
                j = reg.vendedor - 1
                estad[i][j] += reg.monto
                estad[2][j] += reg.monto          # total por vendedor
                estad[i][15] += reg.monto         # total por tipo
                estad[2][15] += reg.monto         # total general
    except FileNotFoundError:
        print("Error al abrir el archivo de entrada...")
        sys.exit(1)
    mostrar_matriz(estad)

if __name__ == "__main__":
    main()